packages/graph/hash_graph/lib/graph/src/shared/identifier/mod.rs

```diff
@@ -13,14 +23,177 @@ pub mod account;
 pub mod knowledge;
 pub mod ontology;
 
-pub type Timestamp = DateTime<Utc>;
+#[derive(
+    Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, ToSql,
+)]
+#[serde(transparent)]
+#[postgres(transparent)]
+pub struct DecisionTimestamp(DateTime<Utc>);
+
+impl fmt::Display for DecisionTimestamp {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt::Display::fmt(&self.0, fmt)
+    }
+}
+
+impl FromStr for DecisionTimestamp {
+    type Err = chrono::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(DateTime::from_str(s)?))
+    }
+}
+
+#[derive(
+    Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, ToSql,
+)]
+#[serde(transparent)]
+#[postgres(transparent)]
+pub struct TransactionTimestamp(DateTime<Utc>);
+
+impl fmt::Display for TransactionTimestamp {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt::Display::fmt(&self.0, fmt)
+    }
+}
+
+impl FromStr for TransactionTimestamp {
+    type Err = chrono::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(DateTime::from_str(s)?))
+    }
+}
 
 #[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
-pub struct Timespan {
-    pub from: Timestamp,
-    pub to: Option<Timestamp>,
+pub(crate) struct Timespan {
+    start: Bound<DateTime<Utc>>,
+    end: Bound<DateTime<Utc>>,
+}
+
+impl FromSql<'_> for Timespan {
+    fn from_sql(_: &Type, buf: &[u8]) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
+        fn parse_bound(
+            bound: &RangeBound<Option<&[u8]>>,
+        ) -> Result<Bound<DateTime<Utc>>, Box<dyn std::error::Error + Send + Sync>> {
+            Ok(match bound {
+                RangeBound::Inclusive(None) | RangeBound::Exclusive(None) => {
+                    unimplemented!("null ranges are not supported")
+                }
+                RangeBound::Inclusive(Some(bytes)) => {
+                    let timestamp = DateTime::from_sql(&Type::TIMESTAMPTZ, bytes)?;
+                    Bound::Included(timestamp)
+                }
+                RangeBound::Exclusive(Some(bytes)) => {
+                    let timestamp = DateTime::from_sql(&Type::TIMESTAMPTZ, bytes)?;
+                    Bound::Excluded(timestamp)
```

Comment:
Quick question (maybe more related to the SQL `tstzrange`). I can see that the bounds are `'[)'` by default, but it's not clear from our usage that we're operating under that assumption. Should we be making that explicit in the SQL functions using `tstzrange`? And additionally, how could we encode that invariant in Rust (should we do that?)

I.e. `tstzrange(now(), 'infinity')` turns into `tstzrange(now(), 'infinity', '[)')` for explicitness